File: src/dotnet/src/Hprose/Client/HproseHttpClient.cs
/**********************************************************\
|                                                          |
|                          hprose                          |
|                                                          |
| Official WebSite: http://www.hprose.com/                 |
|                   http://www.hprose.net/                 |
|                   http://www.hprose.org/                 |
|                                                          |
\**********************************************************/
/**********************************************************\
 *                                                        *
 * HproseHttpClient.cs                                    *
 *                                                        *
 * hprose http client class for C#.                       *
 *                                                        *
 * LastModified: Jan 1, 2014                              *
 * Author: Ma Bingyao <andot@hprose.com>                  *
 *                                                        *
\**********************************************************/
using System;
using System.Collections;
#if !(dotNET10 || dotNET11 || dotNETCF10)
using System.Collections.Generic;
#endif
using System.IO;
#if !(SILVERLIGHT || WINDOWS_PHONE || Core)
using System.IO.Compression;
#elif !(SL2 || Core)
using System.Net.Browser;
#endif
using System.Net;
using Hprose.IO;
#if !(dotNET10 || dotNET11 || dotNETCF10 || dotNETCF20 || SILVERLIGHT || WINDOWS_PHONE || Core)
using System.Security.Cryptography.X509Certificates;
#endif
using System.Threading;

namespace Hprose.Client {
    public class HproseHttpClient : HproseClient {
#if (PocketPC || Smartphone || WindowsCE)
        private static CookieManager cookieManager = new CookieManager();
#elif !SL2
        private static CookieContainer cookieContainer = new CookieContainer();
#endif
        private class HttpClientContext {
            internal HttpWebRequest request;
            internal HttpWebResponse response;
#if !Core
            internal Timer timer;
#endif
            internal HttpClientContext(HttpWebRequest request, HttpWebResponse response) {
                this.request = request;
                this.response = response;
#if !Core
                this.timer = null;
#endif
            }
        }

        private string url = null;
#if !(dotNET10 || dotNET11 || dotNETCF10)
        private Dictionary<string, string> headers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
#elif MONO
        private Hashtable headers = new Hashtable(StringComparer.OrdinalIgnoreCase);
#else
        private Hashtable headers = new Hashtable(new CaseInsensitiveHashCodeProvider(), new CaseInsensitiveComparer());
#endif
#if !Core
		private int timeout = 30000;
#endif
#if !(SILVERLIGHT || WINDOWS_PHONE)
        private ICredentials credentials = null;
#if !Core
        private bool keepAlive = false;
        private int keepAliveTimeout = 300;
        private IWebProxy proxy = null;
#if !dotNETCF10
        private string connectionGroupName = null;
#if !(dotNET10 || dotNET11 || dotNETCF20)
        private X509CertificateCollection clientCertificates = null;
#endif
#endif
#endif
#endif
        public HproseHttpClient()
            : base() {
        }

        public HproseHttpClient(string uri)
            : base(uri) {
        }

        public HproseHttpClient(HproseMode mode)
            : base(mode) {
        }

        public HproseHttpClient(string uri, HproseMode mode)
            : base(uri, mode) {
        }

        public override void UseService(string uri) {
            this.url = uri;
        }

        public void SetHeader(string name, string value) {
            string nl = name.ToLower();
            if (nl != "content-type" &&
                nl != "content-length" &&
                nl != "host") {
                if (value == null) {
                    headers.Remove(name);
                }
                else {
                    headers[name] = value;
                }
            }
        }

        public string GetHeader(string name) {
#if (dotNET10 || dotNET11 || dotNETCF10)
            return (string)headers[name];
#else
            return headers[name];
#endif
        }

#if !Core
        public int Timeout {
            get {
                return timeout;
            }
            set {
                timeout = value;
            }
        }
#endif

#if !(SILVERLIGHT || WINDOWS_PHONE)
        public ICredentials Credentials {
            get {
                return credentials;
            }
            set {
                credentials = value;
            }
        }

#if !Core
        public bool KeepAlive {
            get {
                return keepAlive;
            }
            set {
                keepAlive = value;
            }
        }

        public int KeepAliveTimeout {
            get {
                return keepAliveTimeout;
            }
            set {
                keepAliveTimeout = value;
            }
        }

        public IWebProxy Proxy {
            get {
                return proxy;
            }
            set {
                proxy = value;
            }
        }

#if !dotNETCF10
        public string ConnectionGroupName {
            get {
                return connectionGroupName;
            }
            set {
                connectionGroupName = value;
            }
        }
#if !(dotNET10 || dotNET11 || dotNETCF20)
        public X509CertificateCollection ClientCertificates {
            get {
                return clientCertificates;
            }
            set {
                clientCertificates = value;
            }
        }
#endif
#endif
#endif
#endif

        protected override object GetInvokeContext() {
            Uri uri = new Uri(url);
#if !(SILVERLIGHT || WINDOWS_PHONE) || SL2
            HttpWebRequest request = WebRequest.Create(uri) as HttpWebRequest;
#else
            HttpWebRequest request = WebRequestCreator.ClientHttp.Create(uri) as HttpWebRequest;
#endif
#if !(SILVERLIGHT || WINDOWS_PHONE)
            request.Credentials = credentials;
#if !Core
            request.ServicePoint.ConnectionLimit = Int32.MaxValue;
            request.Timeout = timeout;
            request.SendChunked = false;
#if !(dotNET10 || dotNETCF10)
            request.ReadWriteTimeout = timeout;
#endif
            request.ProtocolVersion = HttpVersion.Version11;
            if (proxy != null) {
                request.Proxy = proxy;
            }
            request.KeepAlive = keepAlive;
            if (keepAlive) {
                request.Headers.Set("Keep-Alive", KeepAliveTimeout.ToString());
            }
#if !dotNETCF10
            request.ConnectionGroupName = connectionGroupName;
#if !(dotNET10 || dotNET11 || dotNETCF20)
            if (clientCertificates != null) {
                request.ClientCertificates = clientCertificates;
            }
#endif
#endif
#endif
#endif
#if (dotNET10 || dotNET11 || dotNETCF10)
            foreach (DictionaryEntry header in headers) {
                request.Headers[(string)header.Key] = (string)header.Value;
            }
#else
            foreach (KeyValuePair<string, string> header in headers) {
                request.Headers[header.Key] = header.Value;
            }
#endif
#if (PocketPC || Smartphone || WindowsCE)
            request.AllowWriteStreamBuffering = true;
            request.Headers["Cookie"] = cookieManager.GetCookie(uri.Host,
                                                                uri.AbsolutePath,
                                                                uri.Scheme == "https");
#elif !SL2
            request.CookieContainer = cookieContainer;
#endif
            return new HttpClientContext(request, null);
        }

        protected override void SendData(Stream ostream, object context, bool success) {
            ostream.Flush();
#if (dotNET10 || dotNET11 || dotNETCF10 || dotNETCF20)
            ostream.Close();
#else
            ostream.Dispose();
#endif
        }

        protected override void EndInvoke(Stream istream, object context, bool success) {
            HttpClientContext clientContext = (HttpClientContext)context;
#if !Core
            if (clientContext.timer != null) {
                clientContext.timer.Dispose();
                clientContext.timer = null;
            }
#endif
#if (dotNET10 || dotNET11 || dotNETCF10 || dotNETCF20)
            istream.Close();
#else
            istream.Dispose();
#endif
#if dotNET45
            clientContext.response.Dispose();
#else
            clientContext.response.Close();
#endif
        }

#if !(SILVERLIGHT || WINDOWS_PHONE || Core)
        protected override Stream GetOutputStream(object context) {
            HttpWebRequest request = ((HttpClientContext)context).request;
            request.Method = "POST";
            request.ContentType = "application/hprose";
            Stream ostream = request.GetRequestStream();
#if !(PocketPC || Smartphone || WindowsCE)
            ostream = new BufferedStream(ostream);
#endif
            return ostream;
        }

        protected override Stream GetInputStream(object context) {
            HttpClientContext httpClientContent = (HttpClientContext)context;
            HttpWebRequest request = httpClientContent.request;
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
#if (PocketPC || Smartphone || WindowsCE)
            cookieManager.SetCookie(response.Headers.GetValues("Set-Cookie"), request.RequestUri.Host);
            cookieManager.SetCookie(response.Headers.GetValues("Set-Cookie2"), request.RequestUri.Host);
#endif
            Stream istream = response.GetResponseStream();
#if !(PocketPC || Smartphone || WindowsCE)
            istream = new BufferedStream(istream);
#endif
            string contentEncoding = response.ContentEncoding.ToLower();
            if (contentEncoding.IndexOf("deflate") > -1) {
                istream = new DeflateStream(istream, CompressionMode.Decompress);
            }
            else if (contentEncoding.IndexOf("gzip") > -1) {
                istream = new GZipStream(istream, CompressionMode.Decompress);
            }
            httpClientContent.response = response;
            return istream;
        }
#endif
#if !Core
        protected void TimeoutHandler(object state) {
            HttpClientContext context = (HttpClientContext)state;
            if (context.response == null) {
                context.request.Abort();
            }
            else {
                context.response.Close();
            }
            context.timer.Dispose();
            context.timer = null;
        }
#endif
        protected override IAsyncResult BeginGetOutputStream(AsyncCallback callback, object context) {
            HttpWebRequest request = ((HttpClientContext)context).request;
            request.Method = "POST";
            request.ContentType = "application/hprose";
#if !Core
            ((HttpClientContext)context).timer = new Timer(new TimerCallback(TimeoutHandler),
                                                           context,
                                                           timeout,
                                                           0);
#endif
            return request.BeginGetRequestStream(callback, context);
        }

        protected override Stream EndGetOutputStream(IAsyncResult asyncResult) {
            HttpClientContext context = (HttpClientContext)asyncResult.AsyncState;
            Stream ostream = context.request.EndGetRequestStream(asyncResult);
#if !(PocketPC || Smartphone || WindowsCE || SILVERLIGHT || WINDOWS_PHONE || Core)
            ostream = new BufferedStream(ostream);
#endif
            return ostream;
        }

        protected override IAsyncResult BeginGetInputStream(AsyncCallback callback, object context) {
            HttpClientContext httpClientContent = (HttpClientContext)context;
            return httpClientContent.request.BeginGetResponse(callback, context);
        }

        protected override Stream EndGetInputStream(IAsyncResult asyncResult) {
            HttpClientContext httpClientContent = (HttpClientContext)asyncResult.AsyncState;
            HttpWebRequest request = httpClientContent.request;
            HttpWebResponse response = (HttpWebResponse)request.EndGetResponse(asyncResult);
#if (PocketPC || Smartphone || WindowsCE)
            cookieManager.SetCookie(response.Headers.GetValues("Set-Cookie"), request.RequestUri.Host);
            cookieManager.SetCookie(response.Headers.GetValues("Set-Cookie2"), request.RequestUri.Host);
#endif
            Stream istream = response.GetResponseStream();
#if !(PocketPC || Smartphone || WindowsCE || SILVERLIGHT || WINDOWS_PHONE || Core)
            istream = new BufferedStream(istream);
#endif
#if !(SILVERLIGHT || WINDOWS_PHONE || Core)
            string contentEncoding = response.ContentEncoding.ToLower();
            if (contentEncoding.IndexOf("deflate") > -1) {
                istream = new DeflateStream(istream, CompressionMode.Decompress);
            }
            else if (contentEncoding.IndexOf("gzip") > -1) {
                istream = new GZipStream(istream, CompressionMode.Decompress);
            }
#endif
            httpClientContent.response = response;
            return istream;
        }
    }
}

File: src/nodejs/hprose/client/HproseHttpClient.js
/**********************************************************\
|                                                          |
|                          hprose                          |
|                                                          |
| Official WebSite: http://www.hprose.com/                 |
|                   http://www.hprose.net/                 |
|                   http://www.hprose.org/                 |
|                                                          |
\**********************************************************/

/**********************************************************\
 *                                                        *
 * HproseHttpClient.js                                    *
 *                                                        *
 * HproseHttpClient for Node.js.                          *
 *                                                        *
 * LastModified: Nov 18, 2012                             *
 * Author: Ma Bingyao <andot@hprose.com>                  *
 *                                                        *
\**********************************************************/

var util = require('util');
var http = require('http');
var https = require('https');
var parse = require('url').parse;

var HproseException = require('../common/HproseException.js');
var HproseClient = require('./HproseClient.js');

var s_cookieManager = {};

function setCookie(headers, host) {
    for (var name in headers) {
        var name = name.toLowerCase();
        var value = headers[name];
        if ((name == 'set-cookie') || (name == 'set-cookie2')) {
            var cookies = value.replace(/(^\s*)|(\s*$)/g, "").split(';');
            var cookie = {};
            value = cookies[0].replace(/(^\s*)|(\s*$)/g, "").split('=', 2);
            if (value[1] === undefined) value[1] = null;
            cookie['name'] = value[0];
            cookie['value'] = value[1];
            for (var i = 1; i < cookies.length; i++) {
                value = cookies[i].replace(/(^\s*)|(\s*$)/g, "").split('=', 2);
                if (value[1] === undefined) value[1] = null;
                cookie[value[0].toUpperCase()] = value[1];
            }
            // Tomcat can return SetCookie2 with path wrapped in "
            if (cookie['PATH']) {
                if (cookie['PATH'].charAt(0) == '"') {
                    cookie['PATH'] = cookie['PATH'].substr(1);
                }
                if (cookie['PATH'].charAt(cookie['PATH'].length - 1) == '"') {
                    cookie['PATH'] = cookie['PATH'].substr(0, cookie['PATH'].length - 1);
                }
            }
            else {
                cookie['PATH'] = '/'
            }
            if (cookie['EXPIRES']) {
                cookie['EXPIRES'] = Date.parse(cookie['EXPIRES']);
            }
            if (cookie['DOMAIN']) {
                cookie['DOMAIN'] = cookie['DOMAIN'].toLowerCase();
            }
            else {
                cookie['DOMAIN'] = host;
            }
            cookie['SECURE'] = (cookie['SECURE'] !== undefined);
            if (s_cookieManager[cookie['DOMAIN']] === undefined) {
                s_cookieManager[cookie['DOMAIN']] = {};
            }
            s_cookieManager[cookie['DOMAIN']][cookie['name']] = cookie;
        }
    }
}

function getCookie(host, path, secure) {
    var cookies = [];
    for (var domain in s_cookieManager) {
        if (host.indexOf(domain) > -1) {
            var names = [];
            for (var name in s_cookieManager[domain]) {
                var cookie = s_cookieManager[domain][name];
                if (cookie['EXPIRES'] && ((new Date()).getTime() > cookie['EXPIRES'])) {
                    names.push(name);
                }
                else if (path.indexOf(cookie['PATH']) === 0) {
                    if (((secure && cookie['SECURE']) ||
                         !cookie['SECURE']) && (cookie['value'] !== null)) {
                        cookies.push(cookie['name'] + '=' + cookie['value']);
                    }
                }
            }
            for (var i in names) {
                delete s_cookieManager[domain][names[i]];
            }
        }
    }
    if (cookies.length > 0) {
        return cookies.join('; ');
    }
    return '';
}

function HproseHttpClient(url) {
    if (this.constructor != HproseHttpClient) return new HproseHttpClient(url);
    HproseClient.call(this);
    var m_options;
    var m_http;
    var m_secure;
    var super_useService = this.useService;
    if (url) useService(url);

    function useService(url) {
        if (url === undefined) return super_useService();
        m_options = parse(url);
        if (m_options.protocol == 'http:') {
            m_http = http;
            m_secure = false;
        }
        else if (m_options.protocol == 'https:') {
            m_http = https;
            m_secure = true;
        }
        else {
            throw new HproseException("Unsupported protocol!");
        }
        m_options.method = "POST";
        m_options.headers = {"connection": "keep-alive"};
        return super_useService();
    }

    function setOption(option, value) {
        if (option != 'method' && option != 'headers') {
            m_options[option] = value;
        }
    }

    function setHeader(name, value) {
        var lname = name.toLowerCase();
        if (lname != 'content-type' &&
            lname != 'host') {
            if (value) {
                m_options.headers[lname] = value;
            }
            else {
                delete m_options.headers[lname];
            }
        }
    }

    this.on('senddata', function(invoker, data) {
        m_options.headers["content-length"] = data.length;
        var cookie = getCookie(m_options.host, m_options.path, m_secure);
        if (cookie != '') {
            m_options.headers['cookie'] = cookie;
        }
        var timeoutid;
        var request = m_http.request(m_options, function(response) {
            var bufferList = [];
            var bufferLength = 0;
            response.on("data", function(chunk) {
                bufferList.push(chunk);
                bufferLength += chunk.length;
            });
            response.on("end", function() {
                if (timeoutid) {
                    clearTimeout(timeoutid);
                    var data = Buffer.concat(bufferList, bufferLength);
                    if (response.statusCode == 200) {
                        invoker.emit('getdata', data);
                    }
                    else {
                        var e = new HproseException(response.statusCode + ':' +  data.toString());
                        invoker.emit('error', e);
                    }
                }
            });
            response.on('error', function(e) {
                clearTimeout(timeoutid);
                invoker.emit('error', e);
            });
            response.on('aborted', function() {
                if(timeoutid) {
                    clearTimeout(timeoutid);
                    invoker.emit('error', new HproseException('response aborted'));
                }
            });
            if (response.statusCode == 200) {
                setCookie(response.headers, m_options.host);
            }
        });

        timeoutid = setTimeout(function() {
            timeoutid = null;
            request.abort();
            invoker.emit('error', new HproseException("timeout"));
        }, this.getTimeout());

        request.on('error', function(e) {
            if (timeoutid) {
                clearTimeout(timeoutid);
                invoker.emit('error', e);
            }
        });
        request.end(data);
    });

    // public methods
    this.useService = useService;
    this.setOption = setOption;
    this.setHeader = setHeader;
}

util.inherits(HproseHttpClient, HproseClient);

module.exports = HproseHttpClient;